### ISSUE TITLE
How can I use deadpool-postgres + diesel?

### ISSUE THREAD

**Issue by NateLing** [bikeshedder/deadpool#104]
I want to user deadpool and deadpool-postgres for diesel, are there any docs for this?

**Comment by bikeshedder** [comment]
Diesel has no `async` support and uses the non-async `postgres` crate.
With `deadpool-postgres` you will get `tokio-postgres` connections which are not useable by `diesel`.

See diesel-rs/diesel#399 and diesel-rs/diesel#2084

In *theory* it is possible to convert a `tokio_postgres::Client` into a `postgres::Client` as the latter is mostly a wrapper for the former. So it could be possible to use `deadpool-postgres`, get a `tokio_postgres::Client`, convert it to a `postgres::Client` and then use it with `diesel`. The actual `diesel` code would still need to run inside a `tokio::spawn_blocking` though as it is not async aware. So I wonder why you would want to use `deadpool-postgres` in the first place.

I feel like there is little to be gained by mixing async `deadpool-postgres` with sync `diesel`.

What is your actual use case and why do you want to do that?

**Comment by weiznich** [comment]
Just to skim in here and clear something up:

> uses the non-async postgres crate.

We don't use the `postgres` crate anywhere, we directly use `libpq` internally.

**Comment by bikeshedder** [comment]
> We don't use the postgres crate anywhere, we directly use libpq internally.

My fault. I could have sworn that I had been using the `postgres` crate last time I tried diesel. I should've checked the `Cargo.toml` of the `diesel` crate once more before making that statement. :see_no_evil: 

I haven't looked into the `Diesel` API for quite some time now. Now I see that the connection handling is outside of diesel and the connection is passed as argument whenever executing a query.

So it should be as simple as writing a `Manager` implementation for `diesel::Connection`. Though there are several stumbling blocks due to `Sync + Send` which is a pain when going from sync to async code and vice versa.

Looking at the `r2d2` implementation for Diesel I can see that it has the same issue:
```rust
unsafe impl<T: Send + 'static> Sync for ConnectionManager<T> {}
```

I do wonder why that's even necessary in the first place and if it's actually correct? :thinking:

**Comment by weiznich** [comment]
See [here](https://github.com/diesel-rs/r2d2-diesel/commit/06fd5606f0e380f1d61eee94a11becedf1a910b1) for the reasoning why this is needed and why this should be correct.

**Comment by bikeshedder** [comment]
I just added some experimental code to support pooling diesel connections in deadpool. I'm still unsure about that unsafe code which marks two structs as Sync, but apart from that it works:

https://github.com/bikeshedder/deadpool/commit/46b5ffa55e923702602533c021b7b7c03d28979c

I just need to implement the `Connection` trait for the wrapper struct so it can be used without that double dereferencing. `&mut **conn` really looks terrible.

**Comment by bikeshedder** [comment]
I just released `deadpool-diesel`: https://crates.io/crates/deadpool-diesel

---

I'm very unsure about these two lines in the code:

https://github.com/bikeshedder/deadpool/blob/master/diesel/src/connection.rs

```rust
unsafe impl<C: diesel::Connection> Sync for Manager<C> {}
```

https://github.com/bikeshedder/deadpool/blob/master/diesel/src/manager.rs

```rust
unsafe impl<C: diesel::Connection + Send + 'static> Sync for ConnectionWrapper<C> {}
```

Maybe someone can enlighten me and help me out with this?

@sgrif You wrote the `r2d2` integration as pointed out by @weiznich. Would you mind to review this regarding the safety concerns?

**Comment by weiznich** [comment]
> I'm very unsure about these two lines in the code:
>
> https://github.com/bikeshedder/deadpool/blob/master/diesel/src/connection.rs
> ```rust
> unsafe impl<C: diesel::Connection> Sync for Manager<C> {}
> ```

This one can be removed by changing the definition of `Manager` to:
```rust
pub struct Manager<C: diesel::Connection> {
    database_url: String,
    _marker: PhantomData<fn() -> C>,
}
```

(See [Subtyping and Variance](https://doc.rust-lang.org/nomicon/subtyping.html) for details about this)

> https://github.com/bikeshedder/deadpool/blob/master/diesel/src/manager.rs
> ```rust
> unsafe impl<C: diesel::Connection + Send + 'static> Sync for ConnectionWrapper<C> {}
> ```

This one is not sound as diesels connections are not `Sync` for various reasons. You cannot just make a wrapper around them `Sync` without using some locking mechanism internally.
That written: It's not clear for me why the `Sync` impl there is even needed. As far as I can see [`Pool` itself is `Sync`](https://docs.rs/deadpool/0.8.1/deadpool/managed/struct.Pool.html#impl-Sync) as long as `Pool::Type` implements `Send`, right? All diesel connections should implement `Send`, but not `Sync`. Therefore I would just drop this line.

**Comment by bikeshedder** [comment]
@weiznich Good catch. I never noticed that `Pool` is only `Sync` if `W` is `Sync` too. That's indeed a mistake and explains all the problems I ran into. I have changed the two uses of `PhantomData<T>` in deadpool to `PhantomData<fn() -> T>` which fixed all compilation errors and enabled me to remove the two wrong and possibly dangerous `unsafe impl`: 4871f984a06a33db24f81a1047001336acbad625

Thanks a lot for pointing that out. :+1:

Now the implementation is `unsafe` free and I'd consider it fairly stable.

@NateLing I've just released `deadpool-diesel 0.1.2` containing all those fixes:

- https://crates.io/crates/deadpool-diesel/0.1.2/

Next step is making it properly compatible with `async-std` as well. For that I created another tracking issue: #106
